lib/AST/ASTScopeCreation.cpp

```diff
@@ -383,21 +383,51 @@ class ScopeCreator final {
 
   void addExprToScopeTree(Expr *expr, ASTScopeImpl *parent) {
     // Use the ASTWalker to find buried captures and closures
-    forEachClosureIn(expr, [&](NullablePtr<CaptureListExpr> captureList,
-                               ClosureExpr *closureExpr) {
-      ifUniqueConstructExpandAndInsert<WholeClosureScope>(parent, closureExpr,
-                                                          captureList);
-    });
+    ASTScopeAssert(expr,
+                 "If looking for closures, must have an expression to search.");
+
+    /// AST walker that finds top-level closures in an expression.
+    class ClosureFinder : public ASTWalker {
+      ScopeCreator &scopeCreator;
+      ASTScopeImpl *parent;
+
+    public:
+      ClosureFinder(ScopeCreator &scopeCreator, ASTScopeImpl *parent)
+          : scopeCreator(scopeCreator), parent(parent) {}
+
+      std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
+        if (auto *closure = dyn_cast<ClosureExpr>(E)) {
+          scopeCreator
+              .ifUniqueConstructExpandAndInsert<ClosureParametersScope>(
+                  parent, closure);
+          return {false, E};
```

Comment:
Before, the code separated out finding the closures from constructing the scopes for them. But here, they are combined. Now, the class name on line 390 isn't quite right. But more to the point, I don't think it serves clarity to combine the enumeration (i.e. finding) with the handling.  I would prefer the `forEachClosureIn` form. When I look at lines 386-390 in the old version, I can see what's going on for each closure without reading through all the code in `ClosureFinder`.

Reply:
I’d have to pass in two lambdas, one to handle capture lists and one to handle closure expressions. It didn’t seem worth it at that point.

Reply:
My quick read didn't catch that. Thanks! So that class would have to be named `BuildScopesForContainedClosuresAndCaptureLists`. And that name still has a bit of a bad smell to my nose. Maybe passing in two lambdas would be better.